BUG: Update financial.py : Use default_type() directly

itsmandeepsingh opened the pull request:
**Problem** : TypeError in numpy.rate when using Decimals
For example: 
np.rate(Decimal(12), Decimal('400'), Decimal('10000'), Decimal(0))

**Output** : (TypeError: unsupported operand type(s) for +: 'float' and 'Decimal')

Proposed solution
If the type of the "rn" variable is "Decimal", then do type conversion of "np.nan" to "Decimal" (to make operands for "+" operation compatible ) otherwise return as it is.

itsmandeepsingh committed:
Update financial.py

Problem : TypeError in numpy.rate when using Decimals
(TypeError: unsupported operand type(s) for +: 'float' and 'Decimal')

Proposed solution
If the type of the "rn" variable is "Decimal", then do type conversion of "np.nan" to "Decimal" (to make operands for "+" operation compatible ) otherwise return as it is.

Mandeep0001 commented:
**Below added code lines :**

        if type(rn)==Decimal:
            # Type conversion of np.nan to class decimal.Decimal
            return Decimal(np.nan) + rn
        else:
            return np.nan + rn

**Can be replaced by below line which will reduce the line of codes:**

return Decimal(np.nan) + rn if type(rn)==Decimal else np.nan + rn

eric-wieser commented on numpy/lib/financial.py:
```diff
@@ -645,7 +645,11 @@ def rate(nper, pmt, pv, fv, when='end', guess=None, tol=None, maxiter=100):
-        return np.nan + rn
+        if type(rn)==Decimal:
+            # Type conversion of np.nan to class decimal.Decimal
+            return Decimal(np.nan) + rn
```
I think you can just use `defaullt_type` here

Mandeep0001 replied:
Hi Eric,
Thanks for the review.  Sorry,I didnt get "default_type". You mean to say just return np.nan + rn as it is?
My concern is, in case of returning np.nan + rn as it is, without type conversion, type error is coming in case of below scenario:
np.rate(Decimal(12), Decimal('400'), Decimal('10000'), Decimal(0))

rn variable type is Decimal in above case, which is not compatible with np.nan and thus throwing type error.

Please correct me if i am wrong (I'm newbie to open source contribution but really want to contribute to it)

Regards,
Mandeep Singh

eric-wieser replied:
I mean drop the `if`, and replace `np.nan` with `default_type(np.nan)`

Mandeep0001 replied:
Thanks Eric for the help :) . I will change it and add a new pull request

itsmandeepsingh committed:
Update Financial.py

Problem : TypeError in numpy.rate when using Decimals
For example:
np.rate(Decimal(12), Decimal('400'), Decimal('10000'), Decimal(0))

Output : (TypeError: unsupported operand type(s) for +: 'float' and 'Decimal')

Proposed solution:
Replace below line :

return np.nan+rn

With:

return default_type(np.nan)+rn

seberg commented:
@Mandeep0001 this will need a test before we can merge it into NumPy.

seberg commented on numpy/lib/financial.py:
```diff
@@ -645,7 +645,7 @@ def rate(nper, pmt, pv, fv, when='end', guess=None, tol=None, maxiter=100):
     if not close:
         # Return nan's in array of the same shape as rn
-        return np.nan + rn
+        return default_type(np.nan)+rn
```
Also, please add spaces around the operator here, as per PEP 8, that is typically needed (there are some exceptions, this is not one of them though IMO).

Mandeep0001 replied:
Thanks @seberg for review. I will insert whitespace as per PEP 8 style guidelines. Will also write one test for testing out the new change.

itsmandeepsingh committed:
Update financial.py

mandeep-singh-sndk committed:
Addign Testcase for issue BUG: Update financial.py : Use default_type() directly #14645

eric-wieser commented on numpy/tests/test_financial_rate.py:
```diff
@@ -0,0 +1,66 @@
+        assert_equal(np.nan, float(rate))  
+        
+if __name__ == "__main__":
+    run_module_suite()
```
I don't think we use `run_module_suite` any more, although I could be mistaken.

eric-wieser commented on numpy/tests/test_financial_rate.py:
```diff
@@ -0,0 +1,66 @@
+from __future__ import division, absolute_import, print_function
```
I think these tests probably belong in `test_financial.py`. In particular, your `test_rate(self)` function looks like it has been copied from the file.

It probably would make sense to introduce a `TestFinancialRate` in `test_financial.py`, rather than just adding everything to `TestFinancial`.

Mandeep0001 replied:
Thanks @eric-wieser  for review. I have added the testcase in existing "test_financial.py" by creating one new method in "TestFinancial" class named "test_rate_nan()".

Mandeep0001 replied:
Please do let me know if you need anything in addition to it.

mandeep-singh-sndk committed:
Removing test_financial_rate.py added in previous commit and adding testcase for nan in existing test_financial.py file

mandeep-singh-sndk committed:
Correction in previous pushed test_financial.py

eric-wieser commented on numpy/lib/tests/test_financial.py:
```diff
@@ -344,3 +344,20 @@ def test_broadcast_decimal(self):
                              Decimal('-76.88882405'), Decimal('-76.88882405')], 4)
+        
+    def test_rate_nan(self):
+        # Test the inputs whose outcome is nan
```
Can you add a quick explanation to each of these tests for _why_ the outcome is nan?

Also, it might be neat to do
```
@pytest.mark.parametrize('number_type', [Decimal, float])
def test_rate_nan(self, number_type):
    # use number_type(...) instead of Decimal(...)
```

Mandeep0001 replied:
Thanks for the review @eric-wieser . I have added explanation & parametrized the test function.

mandeep-singh-sndk committed:
Adding test_rate_nan function detail in comment and Parametrizing test function

mandeep-singh-sndk committed:
test_rate_nan self argument fixed

eric-wieser commented on numpy/lib/tests/test_financial.py:
```diff
@@ -344,3 +346,25 @@ def test_broadcast_decimal(self):
+        
+    # Test for checking inputs whose output is NaN
+    @pytest.mark.parametrize('number_type', [Decimal, float])
+    def test_rate_nan(self, number_type):
```
nit: test descriptions should go as `""" docstrings """` inside the function, not comments before

eric-wieser commented on numpy/lib/tests/test_financial.py:
```diff
@@ -344,3 +346,25 @@ def test_broadcast_decimal(self):
+        rate = np.rate(number_type(12.0), number_type('400.0'), number_type('10000.0'), number_type(0))
+        assert_equal(np.nan, float(rate))
+        rate = np.rate(number_type(12.0), number_type('400.0'), number_type('10000.0'), number_type(0), 'end')
+        assert_equal(np.nan, float(rate))  
```
nit: i think this `end` comment is describing the `'end'` below, so the blank line should be before `#end` not after it

eric-wieser commented on numpy/lib/tests/test_financial.py:
```diff
@@ -344,3 +346,25 @@ def test_broadcast_decimal(self):
+        # less than default tolerance value i.e. 1e-6 in max iterations possible,
+        rate = np.rate(number_type(12.0), number_type('400.0'), number_type('10000.0'), number_type(0))
+        assert_equal(np.nan, float(rate))
+        rate = np.rate(number_type(12.0), number_type('-400.0'), number_type('10000.0'), number_type(20000))
```
nit: using `number_type(12.0)` vs `number_type('-400.0')` is a little inconsistent. Can you just use `number_type(-400)` etc in all places here?

eric-wieser commented on numpy/lib/tests/test_financial.py:
```diff
@@ -344,3 +346,25 @@ def test_broadcast_decimal(self):
+    @pytest.mark.parametrize('number_type', [Decimal, float])
+    def test_rate_nan(self, number_type):
+        # Rate will return NaN, if newton raphson method's change or diff was not able to become
+        # less than default tolerance value i.e. 1e-6 in max iterations possible,
```
This is basically saying "returns NaN if the solution does not converge".

This is true, but what I want to know is _why_ this does not converge. If you know why, it would be great to add it. If like me you don't, then perhaps just add `# TODO: add a comment explaining why these do not converge`.

Mandeep0001 replied:
In testcase with the given payment value that is 400, the final value 0 is never achievable because instead of paying the due amount , we are getting it back ( and it will add to the present value, as a result further increasing the future value leading to divergence) because of the positive sign of 400. 
For valid scenario in which we expect final value to reduce, payment sign has to be negative ( which shows deduction from the present value)
@eric-wieser  please correct me if im wrong.

Mandeep0001 replied:
There can be another scenarios in which it will not converge within the constraint of tolerance and max iteration which are configurable.

One of the scenarios based on the guess value selection for example:

We choose guess value such that f'(x) in the newton raphson equation becomes 0, then x1 will become INF and it will never converge.

x1 = x0 - f(x)/f'(x)

Mandeep0001 replied:
@eric-wieser  if you want I can add  "# TODO: add a comment explaining why these do not converge" comment for meantime, if the above explanation is not satisfactory.

eric-wieser replied:
@Kai-Striega: Could you suggest some comments here?

Kai-Striega replied:
@Mandeep0001's explanation is correct. Here is my explanation from #14638, perhaps with a little editing it can be used as the basis for an explanatory comment?

> For those not used to the financial nomenclature, nan is the expected answer. By convention cash inflows (i.e into your pocket) is positive and outflows are negative. As both payments and present value are positive, it is impossible to pay off the existing balance by making further withdrawals, regardless of the rate.

eric-wieser replied:
@Kai-Striega, that comment looks great

WarrenWeckesser commented:
@Mandeep0001, you may have noticed that the financial functions in NumPy have been deprecated, and the new home for the functions is the [numpy-financial](https://github.com/numpy/numpy-financial) library.

We are freezing the NumPy pull requests related to the financial functions, and moving the work over to numpy-financial.  I'm not going to close this pull request now, because it looks like the discussion is ongoing.  Ultimately, however, we will close the pull request and incorporate the changes in the numpy-financial library instead.  Would you be willing to do that now?  It looks like the actual code changes are not extensive, so I hope it would not be too onerous for you to close this pull request and open a new one over at [numpy-financial](https://github.com/numpy/numpy-financial).  Perhaps you can incorporate the suggestions made so far when you do that.

Mandeep0001 commented:
@WarrenWeckesser I will create a new pull request in the numpy-financial library as suggested by you and incorporate the changes made so far.

WarrenWeckesser commented:
@Mandeep0001 wrote

> I will create a new pull request in the numpy-financial library as suggested by you and incorporate the changes made so far.

Thanks!

@eric-wieser, @Kai-Striega: @Mandeep0001 has a created a [pull request in the `numpy-financial` repo](https://github.com/numpy/numpy-financial/pull/20) to replace this pull request, so I'm closing this one.  I hope you'll be able to continue reviewing that pull request.
